src/cargo/util/toml/mod.rs

```diff
@@ -1944,7 +1980,13 @@ fn detailed_dep_to_dependency<P: ResolveToPath + Clone>(
     }
 
     if let Some(p) = orig.public {
-        cx.features.require(Feature::public_dependency())?;
+        if !cx.config.cli_unstable().public_dependency
+            && !cx.features.require(Feature::public_dependency()).is_ok()
+        {
+            bail!(format!(
+                "Ignoring `public` on dependency {name}.  Pass `-Zpublic-dependency` to enable support for it", name = dep.name_in_toml()
```

Comment:
`read_manifest_from_str` will not step into here, because had set orig.public = None before.

There are calls from other places, but I'm not sure